jolt-core/src/subprotocols/booleanity.rs

```diff
@@ -0,0 +1,596 @@
+use allocative::Allocative;
+#[cfg(feature = "allocative")]
+use allocative::FlameGraphBuilder;
+use ark_std::Zero;
+use rayon::prelude::*;
+use std::{
+    cell::RefCell,
+    ops::{Deref, DerefMut},
+    rc::Rc,
+    sync::Arc,
+};
+
+use crate::{
+    field::{JoltField, MaybeAllocative},
+    poly::{
+        eq_poly::EqPolynomial,
+        multilinear_polynomial::{BindingOrder, PolynomialBinding},
+        opening_proof::{
+            OpeningAccumulator, OpeningPoint, ProverOpeningAccumulator, SumcheckId,
+            VerifierOpeningAccumulator, BIG_ENDIAN,
+        },
+        ra_poly::RaPolynomial,
+        split_eq_poly::GruenSplitEqPolynomial,
+    },
+    transcripts::Transcript,
+    utils::{math::Math, thread::drop_in_background_thread},
+    zkvm::witness::CommittedPolynomial,
+};
+
+use crate::subprotocols::sumcheck::SumcheckInstance;
+
+/// Common prover state for all booleanity sumchecks
+#[derive(Allocative)]
+pub struct BooleanityProverState<F: JoltField> {
+    /// B: split-eq over address-chunk variables (phase 1, LowToHigh).
+    pub B: GruenSplitEqPolynomial<F>,
+    /// D: split-eq over time/cycle variables (phase 2, LowToHigh).
+    pub D: GruenSplitEqPolynomial<F>,
+    /// G as in the Twist and Shout paper
+    pub G: Vec<Vec<F>>,
+    /// H as in the Twist and Shout paper
+    pub H: Vec<RaPolynomial<u8, F>>,
+    /// F: Expanding table
+    pub F: Vec<F>,
+    /// eq_r_r
+    pub eq_r_r: F,
+    /// Indices for H polynomials
+    pub H_indices: Vec<Vec<Option<u8>>>,
+}
+
+pub trait BooleanityConfig {
+    fn d(&self) -> usize;
+
+    fn log_k_chunk(&self) -> usize;
+
+    fn k_chunk(&self) -> usize {
+        1 << self.log_k_chunk()
+    }
+
+    fn log_t(&self) -> usize;
+
+    fn polynomial_type(i: usize) -> CommittedPolynomial;
+
+    fn sumcheck_id() -> SumcheckId;
+}
+
+/// Booleanity Sumcheck interface
+pub trait BooleanitySumcheck<F: JoltField, T: Transcript>:
+    BooleanityConfig + Send + Sync + MaybeAllocative
+{
+    fn gamma(&self) -> &[F::Challenge];
+
+    fn r_address(&self) -> &[F::Challenge];
+
+    fn prover_state(&self) -> Option<&BooleanityProverState<F>>;
+
+    fn prover_state_mut(&mut self) -> Option<&mut BooleanityProverState<F>>;
+
+    fn get_r_cycle(&self, accumulator: &dyn OpeningAccumulator<F>) -> Vec<F::Challenge>;
+
+    fn booleanity_degree(&self) -> usize {
+        3
+    }
+
+    fn booleanity_num_rounds(&self) -> usize {
```

Comment:
```suggestion
    fn num_rounds(&self) -> usize {
```
Given that we are wrapping `B: BooleanitySumcheck` into a `Booleanity` struct, we don't actually need to prefix these methods, since they will not conflict with `SumcheckInstance` at all.
And then we can do `<self as BooleanitySumcheck>::degree()` or so.